[UnifiedPDF] mail.proton.me: Encrypted <object> PDF does not show "document is password protected" UI

aprotyas opened the pull request:
#### 2528cea2fd90d74921dd1fee91edaccef6103ced
<pre>
[UnifiedPDF] mail.proton.me: Encrypted &lt;object&gt; PDF does not show &quot;document is password protected&quot; UI
<a href="https://bugs.webkit.org/show_bug.cgi?id=282113">https://bugs.webkit.org/show_bug.cgi?id=282113</a>
<a href="https://rdar.apple.com/138051102">rdar://138051102</a>

Reviewed by Ryosuke Niwa.

In PDFLayerController, we got a password prompt for free because PDFKit
would paint one into a sublayer of the root document layer. In the
Unified PDF plugin, we implement this password prompt as a &lt;div&gt; element
on the annotation container.

This works great for main frame and &lt;iframe&gt; PDF plugins, since these
plugin types support annotation forms. However, for &lt;embed&gt;/&lt;object&gt;
plugins, since we don&apos;t support annotation forms, we are unable to
present the password prompt &lt;div&gt; and are left with an empty grey view.

To address this missing functionality, we implement the subset of
forms support required to present the password &lt;prompt&gt; div in embedded
plugins. We achieve this by attaching a shadow root under an embedded
plugin, which we then use as the host element to attach the annotation
container/password prompt elements to. More Details about this
implementation provided inline.

* Source/WebCore/dom/Element.h:
  WEBCORE_EXPORT userAgentShadowRoot() so we can call into it from
  WebKit::UnifiedPDFPlugin.

* Source/WebCore/html/HTMLObjectElement.cpp:
(WebCore::HTMLObjectElement::renderFallbackContent):
  Special case the &lt;embed&gt;/&lt;object&gt; PDF case by not defaulting to
  fallback content on what we think may be a failed plugin creation
  attempt. We gate this behavior between presence of a UA shadow root on
  the embedded object, which would indicate we&apos;ve simply invalidated
  renderers and not actually failed to create the requested plugin
  object.

* Source/WebCore/rendering/RenderAttachment.h:
* Source/WebCore/rendering/RenderEmbeddedObject.cpp:
(WebCore::RenderEmbeddedObject::layout):
  Teach RenderEmbeddedObject to also kick off layout for any shadow
  content, if needed.
* Source/WebCore/rendering/RenderEmbeddedObject.h:

* Source/WebCore/rendering/RenderImage.cpp:
(WebCore::RenderImage::layoutShadowContent): Deleted.
  Move this up from RenderImage to RenderReplaced so that
  RenderEmbeddedObject can reuse this shadow content layout logic too.
* Source/WebCore/rendering/RenderImage.h:

* Source/WebCore/rendering/RenderReplaced.cpp:
(WebCore::RenderReplaced::paint):
(WebCore::RenderReplaced::layoutShadowContent):
  Special case the fragemented flow and new/old size comparison logic
  for RenderImage, so we don&apos;t introduce any unexpected behavior change
  during shadow content layout. However, skip this logic elsewhere since
  this seems mostly unnecessary.
* Source/WebCore/rendering/RenderReplaced.h:

* Source/WebKit/WebProcess/Plugins/PDF/PDFPlugin.mm:
(WebKit::PDFPlugin::PDFPlugin):

* Source/WebKit/WebProcess/Plugins/PDF/PDFPluginBase.h:
* Source/WebKit/WebProcess/Plugins/PDF/PDFPluginBase.mm:
(WebKit::PDFPluginBase::supportsForms const):
(WebKit::PDFPluginBase::annotationStyle const):
  Make this a method, rather than a compile-time constant, so we can
  differentiate between the annotation stylesheet for plugins that have
  full support for forms versus those that do not (embedded plugins).
(WebKit::PDFPluginBase::supportsForms): Deleted.
  Make this a const method so it can be called from annotationStyle().

* Source/WebKit/WebProcess/Plugins/PDF/UnifiedPDF/UnifiedPDFPlugin.h:
* Source/WebKit/WebProcess/Plugins/PDF/UnifiedPDF/UnifiedPDFPlugin.mm:
(WebKit::UnifiedPDFPlugin::installAnnotationContainer):
  Introduce a utility method that manages annotation container
  &quot;installation&quot; for the plugin. This method makes sure to instantiate a
  UA shadow root for the requisite plugin types and then appends the
  annotation container div to the appropriate parent element.
(WebKit::UnifiedPDFPlugin::createPasswordEntryForm):
  Enrich the password entry form creation logic by making it aware of
  the UA shadow root we create for certain plugin types. We still gate
  password entry field installation behind full forms support.
(WebKit::UnifiedPDFPlugin::supportsPasswordForm const):

Canonical link: <a href="https://commits.webkit.org/286001@main">https://commits.webkit.org/286001@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/4ac10ee1dad0b08cf49edfaf702e45de4f30ac96

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/74475 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/53904 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/27286 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/78867 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/25713 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/76592 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/63037 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/1689 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/58529 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 win-tests~~](https://ews-build.webkit.org/#/builders/60/builds/16835 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/77542 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/48682 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/64038 "Passed tests") | [  ~~🧪 api-wpe~~](https://ews-build.webkit.org/#/builders/41/builds/38931 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/45723 "Passed tests") | [  ~~🧪 mac-wk1~~](https://ews-build.webkit.org/#/builders/64/builds/21534 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/24046 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/67085 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/21880 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/80382 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/1792 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/122/builds/1047 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/66807 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/1940 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/64056 "Passed tests") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/21/builds/66092 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/10027 "Passed tests") | [❌ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/119/builds/8182 "Failed to checkout and rebase branch from PR 35735") | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/11490 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/1756 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/4544 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/1785 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/1773 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/1792 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

pxlcoder commented on Source/WebKit/WebProcess/Plugins/PDF/UnifiedPDF/UnifiedPDFPlugin.mm:
```diff
@@ -157,17 +159,13 @@ - (void)formChanged:(NSNotification *)notification
+    Ref document = element.document();
+    m_annotationContainer = document->createElement(HTMLNames::divTag, false);
+    m_annotationContainer->setAttributeWithoutSynchronization(HTMLNames::idAttr, "annotationContainer"_s);
+    auto annotationStyleElement = document->createElement(HTMLNames::styleTag, false);
```
```suggestion
    Ref annotationStyleElement = document->createElement(HTMLNames::styleTag, false);
```

pxlcoder commented on Source/WebKit/WebProcess/Plugins/PDF/UnifiedPDF/UnifiedPDFPlugin.mm:
```diff
@@ -348,18 +368,26 @@ static String mutationObserverNotificationString()
-    m_passwordField = passwordField.ptr();
-    passwordField->attach(m_annotationContainer.get());
+    if (supportsForms()) {
+        auto passwordField = PDFPluginPasswordField::create(this);
```
```suggestion
        Ref passwordField = PDFPluginPasswordField::create(this);
```

rr-codes commented on Source/WebKit/WebProcess/Plugins/PDF/UnifiedPDF/UnifiedPDFPlugin.mm:
```diff
@@ -182,6 +180,28 @@ - (void)formChanged:(NSNotification *)notification
+    Ref document = m_element->document();
+
+    if (supportsForms()) {
+        RefPtr { document->bodyOrFrameset() }->appendChild(*m_annotationContainer);
```
won’t this crash in release mode if `m_annotationContainer` is somehow empty?

aprotyas replied:
Correct.. right now it's called exclusively after annotation container setup, but I suppose we should be a bit more robust against the risk of crashing here? Any suggestions?

rr-codes replied:
I usually just do something like 

```
if (!x) {
  ASSERT_NOT_REACHED()
  return
}
```

aprotyas committed:
[UnifiedPDF] mail.proton.me: Encrypted <object> PDF does not show "document is password protected" UI
https://bugs.webkit.org/show_bug.cgi?id=282113
rdar://138051102

Reviewed by Ryosuke Niwa.

In PDFLayerController, we got a password prompt for free because PDFKit
would paint one into a sublayer of the root document layer. In the
Unified PDF plugin, we implement this password prompt as a <div> element
on the annotation container.

This works great for main frame and <iframe> PDF plugins, since these
plugin types support annotation forms. However, for <embed>/<object>
plugins, since we don't support annotation forms, we are unable to
present the password prompt <div> and are left with an empty grey view.

To address this missing functionality, we implement the subset of
forms support required to present the password <prompt> div in embedded
plugins. We achieve this by attaching a shadow root under an embedded
plugin, which we then use as the host element to attach the annotation
container/password prompt elements to. More Details about this
implementation provided inline.

* Source/WebCore/dom/Element.h:
  WEBCORE_EXPORT userAgentShadowRoot() so we can call into it from
  WebKit::UnifiedPDFPlugin.

* Source/WebCore/html/HTMLObjectElement.cpp:
(WebCore::HTMLObjectElement::renderFallbackContent):
  Special case the <embed>/<object> PDF case by not defaulting to
  fallback content on what we think may be a failed plugin creation
  attempt. We gate this behavior between presence of a UA shadow root on
  the embedded object, which would indicate we've simply invalidated
  renderers and not actually failed to create the requested plugin
  object.

* Source/WebCore/rendering/RenderAttachment.h:
* Source/WebCore/rendering/RenderEmbeddedObject.cpp:
(WebCore::RenderEmbeddedObject::layout):
  Teach RenderEmbeddedObject to also kick off layout for any shadow
  content, if needed.
* Source/WebCore/rendering/RenderEmbeddedObject.h:

* Source/WebCore/rendering/RenderImage.cpp:
(WebCore::RenderImage::layoutShadowContent): Deleted.
  Move this up from RenderImage to RenderReplaced so that
  RenderEmbeddedObject can reuse this shadow content layout logic too.
* Source/WebCore/rendering/RenderImage.h:

* Source/WebCore/rendering/RenderReplaced.cpp:
(WebCore::RenderReplaced::paint):
(WebCore::RenderReplaced::layoutShadowContent):
  Special case the fragemented flow and new/old size comparison logic
  for RenderImage, so we don't introduce any unexpected behavior change
  during shadow content layout. However, skip this logic elsewhere since
  this seems mostly unnecessary.
* Source/WebCore/rendering/RenderReplaced.h:

* Source/WebKit/WebProcess/Plugins/PDF/PDFPlugin.mm:
(WebKit::PDFPlugin::PDFPlugin):

* Source/WebKit/WebProcess/Plugins/PDF/PDFPluginBase.h:
* Source/WebKit/WebProcess/Plugins/PDF/PDFPluginBase.mm:
(WebKit::PDFPluginBase::supportsForms const):
(WebKit::PDFPluginBase::annotationStyle const):
  Make this a method, rather than a compile-time constant, so we can
  differentiate between the annotation stylesheet for plugins that have
  full support for forms versus those that do not (embedded plugins).
(WebKit::PDFPluginBase::supportsForms): Deleted.
  Make this a const method so it can be called from annotationStyle().

* Source/WebKit/WebProcess/Plugins/PDF/UnifiedPDF/UnifiedPDFPlugin.h:
* Source/WebKit/WebProcess/Plugins/PDF/UnifiedPDF/UnifiedPDFPlugin.mm:
(WebKit::UnifiedPDFPlugin::installAnnotationContainer):
  Introduce a utility method that manages annotation container
  "installation" for the plugin. This method makes sure to instantiate a
  UA shadow root for the requisite plugin types and then appends the
  annotation container div to the appropriate parent element.
(WebKit::UnifiedPDFPlugin::createPasswordEntryForm):
  Enrich the password entry form creation logic by making it aware of
  the UA shadow root we create for certain plugin types. We still gate
  password entry field installation behind full forms support.
(WebKit::UnifiedPDFPlugin::supportsPasswordForm const):

Canonical link: https://commits.webkit.org/286001@main
